k256/src/arithmetic/scalar/scalar_4x64.rs

```diff
@@ -0,0 +1,459 @@
+
+        let c0 = 0;
+        let c1 = 0;
+        let c2 = 0;
+
+        /* l[0..7] = a[0..3] * b[0..3]. */
+        let (c0, c1) = muladd_fast(self.0[0], rhs.0[0], c0, c1);
+        let (l0, c0, c1) = (c0, c1, 0);
+        let (c0, c1, c2) = muladd(self.0[0], rhs.0[1], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[1], rhs.0[0], c0, c1, c2);
+        let (l1, c0, c1, c2) = (c0, c1, c2, 0);
+        let (c0, c1, c2) = muladd(self.0[0], rhs.0[2], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[1], rhs.0[1], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[2], rhs.0[0], c0, c1, c2);
+        let (l2, c0, c1, c2) = (c0, c1, c2, 0);
+        let (c0, c1, c2) = muladd(self.0[0], rhs.0[3], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[1], rhs.0[2], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[2], rhs.0[1], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[3], rhs.0[0], c0, c1, c2);
+        let (l3, c0, c1, c2) = (c0, c1, c2, 0);
+        let (c0, c1, c2) = muladd(self.0[1], rhs.0[3], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[2], rhs.0[2], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[3], rhs.0[1], c0, c1, c2);
+        let (l4, c0, c1, c2) = (c0, c1, c2, 0);
+        let (c0, c1, c2) = muladd(self.0[2], rhs.0[3], c0, c1, c2);
+        let (c0, c1, c2) = muladd(self.0[3], rhs.0[2], c0, c1, c2);
+        let (l5, c0, c1, _c2) = (c0, c1, c2, 0);
+        let (c0, c1) = muladd_fast(self.0[3], rhs.0[3], c0, c1);
+        let (l6, c0, _c1) = (c0, c1, 0);
+        let l7 = c0;
+
+        WideScalar8x64([l0, l1, l2, l3, l4, l5, l6, l7])
+    }
+
+    pub fn mul(&self, rhs: &Self) -> Self {
+        let wide_res = self.mul_wide(rhs);
+        wide_res.reduce()
+    }
+
+    pub fn from_overflow(w: &[u64; 4], high_bit: Choice) -> Self {
+        let (r2, underflow) = sbb_array_with_underflow(&w, &MODULUS);
+        Self(conditional_select(&w, &r2, !underflow | high_bit))
+    }
+
+    pub fn rshift(&self, shift: usize) -> Self {
+        let full_shifts = shift >> 6;
+        let small_shift = shift & 0x3f;
+
+        let mut res: [u64; 4] = [0, 0, 0, 0];
+
+        if shift > 256 {
+            return Self(res);
+        }
+
+        if small_shift == 0 {
+            #[allow(clippy::needless_range_loop)]
+            #[allow(clippy::manual_memcpy)]
+            for i in 0..(4 - full_shifts) {
+                res[i] = self.0[i + full_shifts];
+            }
+        } else {
+            #[allow(clippy::needless_range_loop)]
+            for i in 0..(4 - full_shifts) {
+                let mut lo = self.0[i + full_shifts] >> small_shift;
+                if i < 3 - full_shifts {
+                    lo |= self.0[i + full_shifts + 1] << (64 - small_shift);
+                }
+                res[i] = lo;
+            }
+        }
+
+        Self(res)
+    }
+
+    #[cfg(feature = "zeroize")]
+    pub fn zeroize(&mut self) {
+        self.0.as_mut().zeroize()
+    }
```

Comment:
Any reason why you're defining `fn zeroize` as an inherent method rather than as `impl Zeroize`? (I see it in a few other places as well, e.g. the 32-bit field and the scalar backends.

Reply:
That's one thing I need to clean up. I was initially confused by all the `add()`/`mul()` etc methods, but it seems that those only exist as inherent methods to simplify defining traits for different combinations of reference/not reference.

Reply:
Fixed.